Add Single Random Draw as an additional coin selection algorithm

achow101 opened the pull request:
To ease in the use of SRD as our fallback mechanism, this PR adds it as a secondary fallback algorithm in addition to the knapsack solver. Since #22009, the solution with the least waste will be chosen. This pattern is continued with SRD simply being another solution whose waste is compared.

achow101 committed:
Add SelectCoinsSRD function

achow101 committed:
tests: wallet_txn explicilty specify inputs

Instead of relying on coin selection to deterministically choose
the correct inputs to use, just specify them explicitly and use
the raw transaction RPCs.

achow101 commented:
Some simulations results: https://gist.github.com/achow101/edf6b5e308035a489fbb1f28d12e2109

I think the important thing to see here is that this is the same or better in every metric except for "mean UTXOs". But I think the thing to note here is that it's about the same mean UTXOs as the positive only effective value simulation. So I think the different there is primarily due to dust outputs that aren't being consumed.

Xekyo commented:
Concept ACK. The results look promising!

> "Better" is defined as having lower fees. If the fees are the same, then the one that chooses more inputs (so consolidates more). If that is also the same, Knapsack is then used (i.e. Knapsack is preferred).

Since you already have a waste metric in BnB, have you considered using the waste metric to pick between the Knapsack and SRD?

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2261,42 +2261,41 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
+    // Filter by the min conf specs and add to utxo_pool and calculate effective value
+    std::vector<OutputGroup> positive_groups;
+    std::vector<OutputGroup> all_groups;
+    for (OutputGroup& group : groups) {
```
The have the same scriptPubKey. IIRC it's for the avoid reuse stuff. Most of the time, they will be single UTXOs.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2261,42 +2261,41 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
-        }
         bnb_used = false;
-        return KnapsackSolver(nTargetValue, utxo_pool, setCoinsRet, nValueRet);
+        return KnapsackSolver(nTargetValue + not_input_fees, all_groups, setCoinsRet, nValueRet);
```
Yes.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2276,27 +2276,14 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
+            if (coin_selection_params.m_subtract_fee_outputs) {
+                group.SetFees(CFeeRate(0), long_term_feerate);
+            } else {
+                group.SetFees(coin_selection_params.effective_fee, long_term_feerate);
```
Yes. It is a `CFeeRate`

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2701,209 +2699,159 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
-                // vouts to the payees
+            // vouts to the payees
+            if (!coin_selection_params.m_subtract_fee_outputs) {
+                coin_selection_params.tx_noinputs_size = 11; // Static vsize overhead + outputs vsize. 4 nVersion, 4 nLocktime, 1 input count, 1 output count, 1 witness overhead (dummy, flag, stack size)
```
This parameter is just generally all of the overhead, regardless of it's location in a transaction. It's in vbytes, which is `weight / 4`

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2701,209 +2699,159 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
+            {
+                CTxOut txout(recipient.nAmount, recipient.scriptPubKey);
+
+                // Include the fee cost for outputs. Note this is only used for BnB right now
```
`tx_noinputs_size` becomes `not_input_fees` later which ultimately increases the selection target so that the selection target covers the fees for the transaction. When subtracting the fee from the outputs, the selection target should not be increasing to cover the fees.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2701,209 +2699,159 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
+            setCoins.clear();
+            int change_spend_size = CalculateMaximumSignedInputSize(change_prototype_txout, this);
+            // If the wallet doesn't know how to sign change output, assume p2sh-p2wpkh
+            // as lower-bound to allow BnB to do it's thing
```
This is moved code, so I think it should remain as is.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2701,209 +2699,159 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
-                        nChangePosInOut = -1;
-                        nFeeRet += nChange;
+                        // Insert change txn at random position:
+                        nChangePosInOut = GetRandInt(txNew.vout.size()+1);
```
No.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2964,7 +2912,7 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
 
     WalletLogPrintf("Fee Calculation: Fee:%d Bytes:%u Needed:%d Tgt:%d (requested %d) Reason:\"%s\" Decay %.5f: Estimation: (%g - %g) %.2f%% %.1f/(%.1f %d mem %.1f out) Fail: (%g - %g) %.2f%% %.1f/(%.1f %d mem %.1f out)\n",
-              nFeeRet, nBytes, nFeeNeeded, feeCalc.returnedTarget, feeCalc.desiredTarget, StringForFeeReason(feeCalc.reason), feeCalc.est.decay,
+              nFeeRet, nBytes, nFeeRet, feeCalc.returnedTarget, feeCalc.desiredTarget, StringForFeeReason(feeCalc.reason), feeCalc.est.decay,
```
`nFeeNeeded` no longer exists. This should have been removed already. I probably forgot to rebase this branch.

achow101 commented:
@Xekyo This PR is built on top of 2 other PRs. It would be better to review those first as many of your comments are for the changes in those PRs. You should start with #17458, then #17331, then this one.

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -329,3 +332,35 @@ bool OutputGroup::EligibleForSpending(const CoinEligibilityFilter& eligibility_f
+    long_term_fee = 0;
+    effective_value = 0;
+    for (CInputCoin& coin : m_outputs) {
+        coin.m_fee = coin.m_input_bytes < 0 ? 0 : effective_feerate.GetFee(coin.m_input_bytes);
```
The default value is -1.

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -329,3 +332,35 @@ bool OutputGroup::EligibleForSpending(const CoinEligibilityFilter& eligibility_f
+    }
+}
+
+OutputGroup OutputGroup::GetPositiveOnlyGroup()
```
How is Positive unspecific? The OutputGroup has only Positive effective values.

Xekyo commented:
@achow101: Okay, I'll take a look at them.

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -167,6 +167,24 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& actual_t
+    out_set.clear();
+    value_ret = 0;
+
+    CAmount selected_value = 0;
```
Consider naming this `selected_eff_value`

achow101 replied:
Done

Xekyo commented on src/wallet/wallet.cpp:
```diff
@@ -2381,9 +2381,62 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
+
+    std::set<CInputCoin> srd_coins;
+    CAmount srd_value;
+    bool srd_ret = SelectCoinsSRD(positive_groups, nTargetValue + change_fee + MIN_FINAL_CHANGE, srd_coins, srd_value);
```
Would be nice if the order of the parameters on `KnapsackSolver` and `SelectCoinsSRD` matched.

achow101 replied:
It would, but changes to `KnapsackSolver` are out of scope for this PR. I prefer to use the current order for `SelectCoinsSRD`.

Xekyo commented on src/wallet/wallet.cpp:
```diff
@@ -2381,9 +2381,62 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
+        setCoinsRet = knapsack_coins;
+        nValueRet = knapsack_value;
+        return true;
+    }
```
How about:

```
if (knapsack_ret) {
        setCoinsRet = knapsack_coins;
        nValueRet = knapsack_value;
}
if (srd_ret) {
    // Use SRD if knapsack had no solution, SRD has lower fees, or SRD has more inputs for same fees
    if (!knapsack_ret || srd_fees < knapsack_fees || srd_fees == knapsack_fees && srd_coins.size() > knapsack_coins.size()) {
        setCoinsRet = srd_coins;
        nValueRet = srd_value;
    }
}
if (knapsack_ret || srd_ret) return true;

// No solution
setCoinsRet.clear();
nValueRet = 0;
return false;
```

achow101 replied:
Done

Xekyo commented:
Concept ACK

jonatack commented:
Concept ACK, will review if #17331 is merged.

glozow commented on src/wallet/wallet.cpp:
```diff
@@ -2393,44 +2394,66 @@ const CTxOut& CWallet::FindNonChangeParentOutput(const CTransaction& tx, int out
-        CAmount cost_of_change = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size) + coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.change_output_size);
+    std::set<CInputCoin> knapsack_coins;
+    CAmount knapsack_value;
+    bool knapsack_ret = KnapsackSolver(nTargetValue + change_fee, all_groups, knapsack_coins, knapsack_value);
```
```suggestion
    const bool knapsack_ret = KnapsackSolver(nTargetValue + change_fee, all_groups, knapsack_coins, knapsack_value);
```

glozow commented on src/wallet/wallet.cpp:
```diff
@@ -2393,44 +2394,66 @@ const CTxOut& CWallet::FindNonChangeParentOutput(const CTransaction& tx, int out
-        std::vector<OutputGroup> groups = GroupOutputs(coins, !coin_selection_params.m_avoid_partial_spends, CFeeRate(0), CFeeRate(0), eligibility_filter, false /* positive_only */);
+    std::set<CInputCoin> srd_coins;
+    CAmount srd_value;
+    bool srd_ret = SelectCoinsSRD(positive_groups, nTargetValue + change_fee + MIN_FINAL_CHANGE, srd_coins, srd_value);
```
```suggestion
    const bool srd_ret = SelectCoinsSRD(positive_groups, nTargetValue + change_fee + MIN_FINAL_CHANGE, srd_coins, srd_value);
```

glozow commented on src/wallet/coinselection.cpp:
```diff
@@ -171,6 +168,24 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& target_v
     return true;
 }
 
+bool SelectCoinsSRD(std::vector<OutputGroup>& utxo_pool, const CAmount& target_value, std::set<CInputCoin>& out_set, CAmount& value_ret)
```
Pretty basic conceptual question, but why only SRD once? Why not run it, say, 100 times, and pick the best one before comparing with knapsack?

Xekyo replied:
That would minimize the input set at high fees and maximize the input set at low fees. If you wanted that, you could directly do largest first selection at high fees and smallest first at low fees. My gut feeling would be that it would grind down the larger UTXOs and very aggressively consolidate your smallest UTXOs at low fees, potentially leading to a large count of similarly sized UTXOs in the mid range. Just picking once with SRD would be expected to give a more even selection of the UTXO pool: if you have a lot of small UTXOs, you'll get more of those, if you only have few and large UTXOs, you'll get a solution with one or few. I.e. one could do that, but it would make the behavior change more extremely respective to the feerate, and I'm not sure whether more extreme is better.

glozow commented on src/wallet/coinselection.cpp:
```diff
@@ -171,6 +168,24 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& target_v
+
+    CAmount selected_eff_value = 0;
+    Shuffle(utxo_pool.begin(), utxo_pool.end(), FastRandomContext());
+    for (const auto& group : utxo_pool) {
```
I know you only pass in positive groups, but I wonder if it could be a good sanity check to add an `Assume(group.effective_value > 0)` here?

achow101 replied:
Done

glozow reviewed (commented):
Concept ACK

In/after the commit _Add SelectCoinsSRD function_, it would be nice to have some unit tests for correctness in coinselector_tests.cpp - gives you the right amount, doesn't always spit out the same result, etc.

jnewbery commented on src/wallet/coinselection.h:
```diff
@@ -102,26 +138,27 @@ struct OutputGroup
-bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret, CAmount not_input_fees);
+bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selection_target, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret);
+
+bool SelectCoinsSRD(std::vector<OutputGroup>& utxo_pool, const CAmount& target_value, std::set<CInputCoin>& out_set, CAmount& value_ret);
```
Please add doxygen comments for new functions.

achow101 replied:
Done

jnewbery commented on src/wallet/coinselection.cpp:
```diff
@@ -171,6 +168,24 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& target_v
+    value_ret = 0;
+
+    CAmount selected_eff_value = 0;
+    Shuffle(utxo_pool.begin(), utxo_pool.end(), FastRandomContext());
```
This calls `std::swap()` up to n-1 times on utxo pool, which seems unnecessary. How about creating a vector of ints from 0 to n-1, shuffling those, and then using that to index into `utxo_pool`. That'd also allow `utxo_pool` to be passed in as a const reference.

achow101 replied:
Done

jnewbery commented on src/wallet/coinselection.cpp:
```diff
@@ -171,6 +168,24 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& target_v
     return true;
 }
 
+bool SelectCoinsSRD(std::vector<OutputGroup>& utxo_pool, const CAmount& target_value, std::set<CInputCoin>& out_set, CAmount& value_ret)
```
Consider returning a `std::optional<std::pair<std::set<CInputCoin>, CAmount>>` rather than using out references and returning a bool to indicate success.

achow101 replied:
Done

jnewbery commented on src/wallet/coinselection.cpp:
```diff
@@ -171,6 +168,24 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& target_v
     return true;
 }
 
+bool SelectCoinsSRD(std::vector<OutputGroup>& utxo_pool, const CAmount& target_value, std::set<CInputCoin>& out_set, CAmount& value_ret)
```
Is there a reason this param is named `utxo_pool` rather than `groups`?

```suggestion
bool SelectCoinsSRD(std::vector<OutputGroup>& groups, const CAmount& target_value, std::set<CInputCoin>& out_set, CAmount& value_ret)
```

achow101 replied:
`utxo_pool` is a more descriptive name.

jnewbery commented on src/wallet/coinselection.cpp:
```diff
@@ -171,6 +168,24 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& target_v
     return true;
 }
 
+bool SelectCoinsSRD(std::vector<OutputGroup>& utxo_pool, const CAmount& target_value, std::set<CInputCoin>& out_set, CAmount& value_ret)
```
Pass simple types like `CAmount` by value:

```suggestion
bool SelectCoinsSRD(std::vector<OutputGroup>& utxo_pool, CAmount target_value, std::set<CInputCoin>& out_set, CAmount& value_ret)
```

achow101 replied:
Done

achow101 commented:
Converting this to a draft as I will be introducing some refactors and other changes that this will be based on.

achow101 committed:
Use SelectCoinsSRD if it has less waste

Try to find a solution with SelectCoinsSRD. If we do have one, add it to
the list of solutions from which we choose the one with the least waste
as the solution to use.

achow101 commented:
I've rebased this onto #22009 because it uses the waste metric to decide which solution to use which I think is a much better and more consistent way than what was implemented here previously.

achow101 committed:
tests: rpc_fundrawtx lock to UTXO types

For some of the tests within rpc_fundrawtx, there is the expectation
that two independent calls to coin selection RPCs will use the same type
of UTXO. This is not necessarily guaranteed, so to make sure it is, use
lockunspent prior to those tests.

achow101 committed:
tests: rpc_fundrawtx use specific inputs for unavailable change test

For the test that checks that there is no error when change is
unavailable but change is also not needed, use specific UTXOs so that
SRD does not cause this to fail when it chooses random inputs.

achow101 committed:
tests: rpc_fundrawtx better test for UTXO inclusion with include_unsafe

Don't assume that specific inputs are going to be used when they aren't
specified explicitly.

Also fixes a bug in the include_unsafe test where after the inputs
confirm, include_unsafe should be set to False rather than True.

achow101 committed:
tests: wallet_basic lock needed unspents

To avoid accidentally spending UTXOs that are needed later in the test,
lock those UTXOs after they're creation.

achow101 commented on src/wallet/coinselection.h:
```diff
@@ -102,26 +138,27 @@ struct OutputGroup
-bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret, CAmount not_input_fees);
+bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selection_target, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret);
+
+bool SelectCoinsSRD(std::vector<OutputGroup>& utxo_pool, const CAmount& target_value, std::set<CInputCoin>& out_set, CAmount& value_ret);
```
Added docs.

naumenkogs commented on src/wallet/coinselection.cpp:
```diff
@@ -168,6 +170,30 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selectio
+    for (const size_t i : indexes) {
+        const OutputGroup& group = utxo_pool.at(i);
+        Assume(group.GetSelectionAmount() == 0);
+        selected_eff_value += group.GetSelectionAmount();
```
sorry, I can't wrap my head around it. If it's assumed to be 0 (prev line), what's the point of adding it and checking it below?

achow101 replied:
Oops, it's supposed to be `>`. Fixed.

Xekyo commented:
> I've rebased this onto #22009 because it uses the waste metric to decide which solution to use which I think is a much better and more consistent way than what was implemented here previously.

In the opening comment of the PR, you mention that BnB is always preferred and if there is no BnB solution, the better of Knapsack or SRD is picked.

Shouldn't all three selection results simply be compared on basis of the waste metric rather than always preferring the BnB solution when one is found? If there is a 5-input BnB solution, and a single input SRD solution at 130 sat/vB, I think there would be an argument to be made to prefer the latter. If that's already the case, please update the OP.

achow101 commented:
Updated the OP

Xekyo commented on src/wallet/coinselection.h:
```diff
@@ -183,6 +185,15 @@ struct OutputGroup
 
 bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selection_target, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret);
 
+/** Select coisn by Single Random Draw. OutputGroups are selected randomly from the eligible
```
```suggestion
/** Select coins by Single Random Draw. OutputGroups are selected randomly from the eligible
```

achow101 replied:
Done

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -168,6 +170,30 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selectio
+    std::set<CInputCoin> out_set;
+    CAmount value_ret = 0;
+
+    std::vector<size_t> indexes;
+    indexes.resize(utxo_pool.size());
+    std::iota(indexes.begin(), indexes.end(), 0);
+    Shuffle(indexes.begin(), indexes.end(), FastRandomContext());
```
Wouldn't it be easier to shuffle the `utxo_pool` directly? Or is this the same object that is used for the other selection algorithms?

achow101 replied:
This change was suggested in https://github.com/bitcoin/bitcoin/pull/17526#discussion_r625212261

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -168,6 +170,30 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selectio
+        value_ret += group.m_value;
+        util::insert(out_set, group.m_outputs);
+        if (selected_eff_value >= target_value) {
+            return std::make_pair(out_set, value_ret);
```
I saw that a selection result object is being introduced in https://github.com/bitcoin/bitcoin/pull/22019. Are you planning for 22019 to succeed or precede this PR?

achow101 replied:
It doesn't particularly matter. Currently I would like for this PR to precede 22019

Xekyo replied:
Okay, thanks

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -382,6 +382,15 @@ bool CWallet::AttemptSelection(const CAmount& nTargetValue, const CoinEligibilit
     }
 
+    // We include the minimum final change for SRD as we do want to avoid making really small change.
+    // KnapsackSolver does not need this because it includes MIN_CHANGE internally.
```
I would prefer if all three selection methods would receive the actual `selection_target` and they would internally add `MIN_FINAL_CHANGE` where applicable. Whether or not an input parameter needs to be amended by a global constant makes more sense as an implementation detail of the called function than each callsite needing to be aware of it, and it would allow the function calls to the selection methods more alike.

I.e. I think adding the `MIN_FINAL_CHANGE` should happen inside of `SelectCoinsSRD(...)`

achow101 replied:
Currently that is difficult to do with the way that we do the waste calculation. I think this can be done in the future, especially post #22019

instagibbs commented:
seems like "fallback" is now a misnomer?

Any simulations/results/intuitions on how often each of the algorithms are likely to be picked?

instagibbs commented on src/wallet/spend.cpp:
```diff
@@ -382,6 +382,15 @@ bool CWallet::AttemptSelection(const CAmount& nTargetValue, const CoinEligibilit
 
+    // We include the minimum final change for SRD as we do want to avoid making really small change.
+    // KnapsackSolver does not need this because it includes MIN_CHANGE internally.
+    const CAmount srd_target = nTargetValue + coin_selection_params.m_change_fee + MIN_FINAL_CHANGE;
```
any simulations/intuitions on what these change outputs look like vs knapsack values, aside from the difference in constant?

Why was this constant `MIN_FINAL_CHANGE` defined but never used before? How was it picked?

achow101 replied:
It looks like we lost it during the effective value change. `MIN_FINAL_CHANGE` used to be the minimum allowed change output with the looping behavior.

instagibbs commented:
concept ACK, I really like the waste heuristic as a picker if nothing else to dislodge bikeshedding about coin selection hierarchies. I'd like to see how much this changes the frequency of change-less solutions in simulations since BnB isn't preferred anymore, since I believe the argument is change-less solutions are simply only preferred after this in an economic sense, not privacy/ending the chaining of change way.

achow101 commented:
> seems like "fallback" is now a misnomer?

Updated the title.

> Any simulations/results/intuitions on how often each of the algorithms are likely to be picked?

Not currently. Perhaps @Xekyo has something?

Xekyo commented:
> > Any simulations/results/intuitions on how often each of the algorithms are likely to be picked?
> 
> Not currently. Perhaps @Xekyo has something?

That depends heavily on the amount of UTXOs that are available and how diverse the UTXO values are. I've seen an exchange wallet do 93% BnB transactions across a timeframe of four weeks just after enabling BnB while they were chewing through a large UTXO pool. The same wallet later settled around 60-70%, IIRC. I'd guess most enterprises with a send-and-receive hot wallet would probably fall into the range of 30-60%, send-only enterprise wallets probably more like 15-30%. Either figure should increase if they're batching payments since BnB solutions are easier to find for larger amounts. 

On an end-user wallet I would expect maybe 5-20% BnB solutions.

Xekyo commented:
> since I believe the argument is change-less solutions are simply only preferred after this in an economic sense, not privacy/ending the chaining of change way.

Given that the change output has a big negative impact on the waste score, the BnB solution should always be preferred, if it has the same count of inputs (given the same type). Due to usually needing to combine multiple inputs in order to arrive at a specific amount and the deterministic search path going from highest effective value to lowest, BnB solutions tend to have a small count of 2 or more inputs. 

Especially in a cluttered wallet, SRD or Knapsack often find a larger solution that may be preferred at low feerates even when it produces a change output. At higher feerates, especially SRD can find the occasional lucky single input solutions that would might have a lower waste score. This may result in slightly fewer BnB solutions than before, but only in cases where the BnB solution was less cost effective.

Given the high estimated cost of change outputs, I'd say that we already have a decent handicap on solutions producing change. It would be possible to further dissuade change, but I'd consider it an open question how to pick the "correct" handicap.

achow101 commented:
I've ran simulations (using the same data) on a recent commit of master as well as on this PR. I've uploaded the results to a gist [for master](https://gist.github.com/achow101/edf6b5e308035a489fbb1f28d12e2109#file-results-0053a5fc7d912d597cd6dc7376b479420d1eae1c0-md) and [for this PR](https://gist.github.com/achow101/edf6b5e308035a489fbb1f28d12e2109#file-results-srd-3eb40005f373032a4d877ffdcc33c15baf97958c-md).

The main result from this particular simulation is that with SRD, less inputs are being consumed overall (mean # of utxos is higher), but fees are slightly lower. SRD ended up being used ~1/3 of the time, BnB ~1/5, and Knapsack the rest.

Xekyo commented:
> I've ran simulations (using the same data) on a recent commit of master as well as on this PR. I've uploaded the results to a gist [for master](https://gist.github.com/achow101/edf6b5e308035a489fbb1f28d12e2109#file-results-0053a5fc7d912d597cd6dc7376b479420d1eae1c0-md) and [for this PR](https://gist.github.com/achow101/edf6b5e308035a489fbb1f28d12e2109#file-results-srd-3eb40005f373032a4d877ffdcc33c15baf97958c-md).
> 
> The main result from this particular simulation is that with SRD, less inputs are being consumed overall (mean # of utxos is higher), but fees are slightly lower. SRD ended up being used ~1/3 of the time, BnB ~1/5, and Knapsack the rest.

Let me try to cast the results slightly differently: the variant with SRD reduces fees by 6% and while that variant spends 296 fewer inputs those seem to be largely accounted for by it creating 255 fewer change outputs in the first place. So yes, the mean UTXO pool is a bit larger, but in effect it also creates 4.4% fewer change outputs in the first place. Not creating so many UTXOs in the first place seems great to me!

achow101 commented:
I redid the simulations with some fixes to how data is gathered and gathered a lot more data. The original simulation was accidentally incorrectly counting usage due to the double selection that occurs for the optimistic avoidpartialspends. I am now also measuing how many uneconomical UTXOs are being spent, and how many changeless txs of each algorithm produces. The aforementioned gist has been updated with the new measurements.

With the new measuring, we see that SRD produces a lot more BnB solutions and ends up spending almost no uneconomical outputs (which would explain the significant increase in UTXOs in the wallet).

glozow reviewed (commented):
code review ACK 981b9d13bdb1743710ea3d0c67ec13ab40619256

I wanted to see if there were any other tests that could be broken due to relying on deterministic coin selection behavior, so I edited `AttemptSelection` to always return the SRD solution; all the functional tests passed.

achow101 commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -47,7 +47,40 @@ def setup_network(self):
         self.connect_nodes(0, 2)
         self.connect_nodes(0, 3)
 
+    def lock_outputs_type(self, wallet, outputtype):
```
No, it's (supposed to be) locking everything of the given type.

glozow replied:
woops :scream:

achow101 commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -944,31 +992,31 @@ def test_include_unsafe(self):
         # And we can also use them once they're confirmed.
         self.nodes[0].generate(1)
-        rawtx = wallet.createrawtransaction([], [{self.nodes[2].getnewaddress(): 3}])
         fundedtx = wallet.fundrawtransaction(rawtx, {"include_unsafe": True})
```
Yes.

achow101 replied:
I've fixed this issue while making changes in the vicinity.

achow101 commented on src/wallet/coinselection.h:
```diff
@@ -183,6 +185,15 @@ struct OutputGroup
+/** Select coins by Single Random Draw. OutputGroups are selected randomly from the eligible
+ * outputs until the target is satisfied
+ *
+ * @param[in]  utxo_pool    The OutputGroups eligible for selection
```
Done

achow101 commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -992,31 +992,31 @@ def test_include_unsafe(self):
         fundedtx = wallet.fundrawtransaction(rawtx, {"include_unsafe": True})
         tx_dec = wallet.decoderawtransaction(fundedtx['hex'])
-        assert any([txin['txid'] == txid1 and txin['vout'] == vout1 for txin in tx_dec['vin']])
+        assert all(any(txin["txid"] == inp[0] and txin["vout"] == inp[1] for inp in inputs) for txin in tx_dec["vin"])
```
Done

glozow commented:
reACK 3633b66 via `git range-diff  981b9d1...3633b66`, thanks for taking the suggestions

meshcollider commented on src/wallet/coinselection.cpp:
```diff
@@ -168,6 +170,30 @@ bool SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool, const CAmount& selectio
+    std::set<CInputCoin> out_set;
+    CAmount value_ret = 0;
+
+    std::vector<size_t> indexes;
```
nit: indices > indexes :)

meshcollider reviewed (commented):
Code review ACK 8bf789b4b4b26082aea1d91c4d7aa8b01aedfdcf

I like that this coin selection algorithm is only 20 lines 😄

laanwj commented:
Concept and code review ACK 3633b667ffca5a715d9fb27e977515c1e24f600a

achow101 commented:
> Does this cause the intermittent wallet_basic issues?

Potentially. The main thing is that this change makes coin selection non-deterministic which means that tests which rely on assuming certain UTXOs are being selected are no longer correct. I thought I got all of the tests that did that, but it is possible that a few were missed.

S3RK commented:
Post merge ACK 3633b66

Reviewed code and independently reproduced simulation results. Confirmed that adding SRD increases amount of changeless solutions.

MarcoFalke commented:
I think this picked up https://github.com/bitcoin/bitcoin/pull/13307#event-6281610244, so I removed the "up for grabs" label there.
